py/stream.c

```diff
@@ -93,13 +93,63 @@ mp_uint_t mp_stream_rw(mp_obj_t stream, void *buf_, mp_uint_t size, int *errcode
     return done;
 }
 
+STATIC mp_uint_t user_read(mp_obj_t obj, void *buf, mp_uint_t size, int *errcode) {
+    mp_obj_t dest[3];
+    mp_load_method(obj, MP_QSTR_read, dest);
+    dest[2] = mp_obj_new_bytearray_by_ref(size, buf);
+    mp_int_t ret = mp_obj_get_int(mp_call_method_n_kw(1, 0, dest));
+    if (ret >= 0) {
+        return ret;
+    } else {
+        *errcode = -ret;
+        return MP_STREAM_ERROR;
+    }
+}
+
+STATIC mp_uint_t user_write(mp_obj_t obj, const void *buf, mp_uint_t size, int *errcode) {
+    mp_obj_t dest[3];
+    mp_load_method(obj, MP_QSTR_write, dest);
+    dest[2] = mp_obj_new_bytearray_by_ref(size, (void*)buf);
+    mp_int_t ret = mp_obj_get_int(mp_call_method_n_kw(1, 0, dest));
+    if (ret >= 0) {
+        return ret;
+    } else {
+        *errcode = -ret;
+        return MP_STREAM_ERROR;
+    }
+}
+
+STATIC mp_uint_t user_ioctl(mp_obj_t obj, mp_uint_t request, uintptr_t arg, int *errcode) {
+    mp_obj_t dest[4];
+    mp_load_method(obj, MP_QSTR_ioctl, dest);
+    dest[2] = mp_obj_new_int(request);
+    dest[3] = mp_obj_new_int_from_uint(arg);
+    mp_int_t ret = mp_obj_get_int(mp_call_method_n_kw(2, 0, dest));
+    if (ret >= 0) {
+        return ret;
+    } else {
+        *errcode = -ret;
+        return MP_STREAM_ERROR;
+    }
+}
+
+STATIC mp_stream_p_t user_stream_p = {
+    user_read,
+    user_write,
+    user_ioctl,
+};
+
 const mp_stream_p_t *mp_get_stream_raise(mp_obj_t self_in, int flags) {
     mp_obj_type_t *type = mp_obj_get_type(self_in);
     const mp_stream_p_t *stream_p = type->protocol;
     if (stream_p == NULL
         || ((flags & MP_STREAM_OP_READ) && stream_p->read == NULL)
         || ((flags & MP_STREAM_OP_WRITE) && stream_p->write == NULL)
         || ((flags & MP_STREAM_OP_IOCTL) && stream_p->ioctl == NULL)) {
+        // if the object is a user instance then return a protocol adaptor
+        if (mp_obj_is_instance_type(type)) {
```

Comment:
>> Your patch adds requirement for (another) check. Where there were no checks at all, t

> No it doesn't. Notice that I removed a check (call to mp_get_stream_raise) so that it balances out.

Here's the check that you added, bingo. Yes, you removed a check somewhere else, but counting it as a "balance" is a way of cheating ;-).